### PR TITLE
One-to-one chat

### PR DESCRIPTION
**Pull request by AndrePu**
**Detailed information:**
#85 One-to-one chats

08/15/19
Modal window for creating new one-to-one dialogs is ready (frontend).

08/23/19
One-to-one chat is ready. It concludes:
- signalR both list of chats and list of messages;
- list of chats: each item of which contains photo, nickname, last message. Length of each text field is checking every time we load item info.
- list of messages: just messages list with photo of sender and sender's nickname. Input field.
- modal chat box: temporary solution for adding personal chats. It contains input for nickname, area-box for message. Handled most part of tricky situations.

On backend:
- Added logic for getting chats by userId;
- Added logic for getting messages by chatId
- Creating chat/message

### CODE REVIEW

File: Cooper.Controllers/ChatController.cs [hunk 1]
```diff
@@ -1,68 +1,111 @@
...
         }
 
-        // GET api/<controller>/5
-        [HttpGet("{id}")]
+        [Authorize]
+        [HttpGet("one-to-one-chats")]
         [ProducesResponseType(200, Type = typeof(Chat))]
         [ProducesResponseType(404)]
-        public IActionResult GetChatById(long id)
+        public IActionResult GetChatByUserId()
```
Comment (horba) [inline Cooper.Controllers/ChatController.cs]: the naming is confusing: if you have in name "ById" it is expected that you pass id and get user with this id

File: Cooper.Controllers/ChatController.cs [hunk 2]
```diff
@@ -1,68 +1,111 @@
...
+
+            if (user == null)
+            {
+                return StatusCode(500); //  connection to database failed
+            }
+
+            IList<Chat> chats = chatRepository.GetPersonalChatsByUserId(user.Id);
 
-            if (chat == null)
+            if (chats == null)
```
Comment (horba) [inline Cooper.Controllers/ChatController.cs:55]: can be length = 0 ?

File: Cooper.Controllers/ChatController.cs [hunk 3]
```diff
@@ -1,68 +1,111 @@
...
+            return Ok(chats);
         }
 
-        // POST api/<controller>
-        [HttpPost]
-        [ProducesResponseType(200)]
-        [ProducesResponseType(201)]
-        public IActionResult Post([FromBody]Chat chat)
+
+        public class Body
```
Comment (horba) [inline Cooper.Controllers/ChatController.cs]: why do you need this?

Comment (horba) [reply Cooper.Controllers/ChatController.cs]: it'better to create separate model and use it

File: Cooper.Controllers/ChatController.cs [hunk 4]
```diff
@@ -1,68 +1,111 @@
...
-                chat.Id = id;
+                Chat chat = chatRepository.GetOnetoOneChatByParticipants(participants);
+
+                if (chat == null)
+                {
+                    chat = new Chat() { Participants = participants };
+                    chat.Id = chatRepository.Create(chat);
+
+                    message.ChatId = chat.Id;
+                    message.Id = messageRepository.Create(message);
```
Comment (horba) [inline Cooper.Controllers/ChatController.cs:88]: it seems like logic for service but ok

File: Cooper/ClientApp/src.v2/app/components/chat-box/chat-box.component.html [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      </mat-list-item>
+
+    </mat-list>
+  </div>
+  
+    <div class="new-message-container">
+
+      <div class="input-container">
+        <textarea id='box' #box placeholder="Enter your message..."
+        class="text-area" (keyup.enter)="createMessage(box.value)"></textarea>
```
Comment (horba) [inline Cooper/ClientApp/src.v2/app/components/chat-box/chat-box.component.html]: why not via ngModel?

File: Cooper/ClientApp/src.v2/app/components/chat-box/chat-box.component.ts [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+  public messageContent: string;
+
+  public emptyMessageFieldError: boolean = false;
+  public emptyMessageErrorContent: string = 'Please, enter your message first';
+
+  public emptyNicknameFieldError: boolean = false;
+  public emptyNicknameErrorContent: string = 'Please, point message recipient before sending a message';
+
+  public userNotFoundErrorContent: string = 'User is not found';
```
Comment (horba) [inline Cooper/ClientApp/src.v2/app/components/chat-box/chat-box.component.ts]: all messages via translates

File: Cooper/ClientApp/src.v2/app/components/chat-box/chat-box.component.ts [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    for (const user of this.chat.participants) {
+      if (user.id === message.senderId) {
+        return user.photoURL;
+      }
+    }
+
+    return '';
+  }
+
+  public getMessageDate(message: Message): string {
```
Comment (horba) [inline Cooper/ClientApp/src.v2/app/components/chat-box/chat-box.component.ts]: it's better to use or create pipe

File: Cooper/ClientApp/src.v2/app/components/chats-list/chats-list.component.ts [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+  public getChatImage(chat: Chat): string {
+
+      return (this.currentSessionUserId === chat.participants[0].id) ? chat.participants[1].photoURL : chat.participants[0].photoURL;
+
+  }
+
+  public getChatName(chat: Chat): string {
+
+    let chatHeader = (this.currentSessionUserId === chat.participants[0].id) ?
```
Comment (horba) [inline Cooper/ClientApp/src.v2/app/components/chats-list/chats-list.component.ts]: might, it should be on backend side?

File: Cooper/ClientApp/src.v2/app/components/chats-list/chats-list.component.ts [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+  public onChatItemClick(chat: Chat): void {
+
+    this.closeNewMessageBlock.emit();
+
+    this.loadChat.emit(chat);
+  }
+
+  public getChatImage(chat: Chat): string {
+
+      return (this.currentSessionUserId === chat.participants[0].id) ? chat.participants[1].photoURL : chat.participants[0].photoURL;
```
Comment (horba) [inline Cooper/ClientApp/src.v2/app/components/chats-list/chats-list.component.ts]: backend side

File: Cooper/ClientApp/src.v2/app/components/chats-list/chats-list.component.ts [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+    if (messageContent.length > fieldPermissibleMaxLength) {
+      messageContent = messageContent.slice(0, fieldPermissibleMaxLength - 3) + '...';
+      return messageContent;
+    } else {
+      return messageContent;
+    }
+  }
+
+  public getLastMessageSentTime(chat: Chat): string {
```
Comment (horba) [inline Cooper/ClientApp/src.v2/app/components/chats-list/chats-list.component.ts]: pipe

**Review by horba — CHANGES_REQUESTED** [review]
- problems of DTO: a lot of model changes at frontend side which can be done at server side

- our dummy component is sensible of data model changes

- for presentantional changes of data use pipes

- small changes

File: Cooper.Controllers/ChatController.cs (hunk 2 above)
Comment (AndrePu) [reply Cooper.Controllers/ChatController.cs:55]: No. We initialize list only when we have at least one record got from query to database

File: Cooper/ClientApp/src.v2/app/components/chat-box/chat-box.component.ts (hunk 6 above)
Comment (AndrePu) [reply Cooper/ClientApp/src.v2/app/components/chat-box/chat-box.component.ts]: I'll do it in separate task where I'll add translations for registration, my-profile, home-page, and chats accordingly

File: Cooper.Controllers/ChatController.cs (hunk 4 above)
Comment (AndrePu) [reply Cooper.Controllers/ChatController.cs:88]: Just wanted to keep it simple =)

File: Cooper/ClientApp/src.v2/app/components/chat-box/chat-box.component.ts (hunk 6 above)
Comment (AndrePu) [reply Cooper/ClientApp/src.v2/app/components/chat-box/chat-box.component.ts]: However it's not big deal to make it now

File: Cooper.Controllers/ChatController.cs (hunk 4 above)
Comment (horba) [reply Cooper.Controllers/ChatController.cs:88]: but it makes it harder. but see in future ^)

File: Cooper/ClientApp/src.v2/app/components/chat-box/chat-box.component.ts [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+      setTimeout(() => {
+        if (!this.userNotFoundError) {
+          this.username = '';
+        }
+      }, 2000);
+
+    } else {
+
+      message.content = this.messageContent.trimRight();
```
Comment (horba) [inline Cooper/ClientApp/src.v2/app/components/chat-box/chat-box.component.ts]: pipe

File: Cooper/ClientApp/src.v2/app/components/chats-list/chats-list.component.html [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+                  <span>{{getChatName(chat)}}</span>
+                </div>
+
+                <div class="date column">
+                    <span align="right">{{chat.messages[chat.messages.length - 1].createDate | messageDateTime}}</span>
+                </div>
+              </div>
+              
+              <div class="row chat-last-message">
+                <i *ngIf="chat && currentSessionUserId===chat.messages[chat.messages.length-1].senderId">{{'CHATS.CHAT-LIST.MESSAGE-PREFIX'|translate}}: &nbsp;</i>
```
Comment (horba) [inline Cooper/ClientApp/src.v2/app/components/chats-list/chats-list.component.html]: too long

File: Cooper.DAO/ChatDAO/ChatDAO.cs [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+        public IList<ChatDb> GetPersonalChats(long userId)
+        {
+            IList<ChatDb> personalChats = null;
+
+            var chatAttributes = new HashSet<string>()
+            {
+                "ID"
+            };
+
+            string sqlExpression = String.Format("SELECT IDCHAT as ID FROM {0} INNER JOIN USERSCHATS ON {0}.ID = USERSCHATS.IDCHAT WHERE IDUSER = {1}", table, userId);
```
Comment (Hunroll) [inline Cooper.DAO/ChatDAO/ChatDAO.cs]: Please, check performance of this query and something like this:
select IDCHAT as ID from {0} where exists (select 1 from USERSCHATS where {0}.ID = USERSCHATS.IDCHAT AND IDUSER = {1}

You can use "Explain plan" in sql developerr

File: Cooper.DAO/ChatDAO/ChatDAO.cs [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+
+            return personalChats;
+
+        }
+
+        public ChatDb GetOnetoOneChatByParticipantsId(IList<long> participantsId)
+        {
+            ChatDb chat = null;
+
+            string sqlExpression = "SELECT chat.id, chat.chatname, chat.photourl FROM CHATS chat " +
```
Comment (Hunroll) [inline Cooper.DAO/ChatDAO/ChatDAO.cs]: same story here. No need to join tables if you select only columns from one. Please, compare performance. If is's the same, then it's OK.

File: Cooper.DAO/MessageDAO/MessageDAO.cs [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+                    messages.Add(message);
+                }
+            }
+
+            return messages;
+        }
+
+        public bool ReadNewMessages(IList<long> messages)
+        {
+            string sqlExpression = $"UPDATE {table} SET ISREAD=\'y\' ";
```
Comment (Hunroll) [inline Cooper.DAO/MessageDAO/MessageDAO.cs]: Can't DBTools generete this query with "IN"?

File: Cooper.DAO/MessageDAO/MessageDAO.cs [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+
+        private IList<EntityORM> Get(string table, HashSet<string> attributes, Dictionary<string, object> where_attributes)
+        {
+            IList<EntityORM> entities = null;
+
+            try
+            {
+                dbConnect.OpenConnection();
+
+                string where_part = string.Join(" AND ", where_attributes.Select(x => string.Format("{0}={1}", x.Key, x.Value)));
```
Comment (Hunroll) [inline Cooper.DAO/MessageDAO/MessageDAO.cs]: DBTools

**Review by Hunroll — CHANGES_REQUESTED** [review]

File: Cooper.DAO/MessageDAO/MessageDAO.cs (hunk 15 above)
Comment (AndrePu) [reply Cooper.DAO/MessageDAO/MessageDAO.cs]: No. Logic for generating update query for more than one entities is abcent as well as logic in CRUD for updating several entities

File: Cooper.DAO/ChatDAO/ChatDAO.cs (hunk 14 above)
Comment (AndrePu) [reply Cooper.DAO/ChatDAO/ChatDAO.cs]: But I need to get all the chats in which all the users in list are in. For this I have only users ID's. So I need to use "USERSCHATS" table to get records from "CHATS" table.

Comment (AndrePu) [reply Cooper.DAO/ChatDAO/ChatDAO.cs]: It's still the same but without JOIN query is a bit more appropriate

File: Cooper.DAO/ChatDAO/ChatDAO.cs [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+
+                    personalChats.Add(chat);
+
+                    chatsIndexes.Add(chat.Id, i);
+                }
+
+                var user_attribute = new HashSet<string> { "IDUSER", "IDCHAT" };
+                sqlExpression = "SELECT IDUSER, IDCHAT FROM USERSCHATS ";
+
+                for (int i = 0; i < personalChats.Count; i++)
```
Comment (Hunroll) [inline Cooper.DAO/ChatDAO/ChatDAO.cs]: WhereRequest.Operators.IN

File: Cooper.DAO/ChatDAO/ChatDAO.cs [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+        {
+            ChatDb chat = null;
+
+            string sqlExpression = "SELECT ID, CHATNAME, PHOTOURL FROM CHATS " +
+                                    "WHERE EXISTS " +
+                                    "(SELECT record1.idchat, record1.iduser FROM USERSCHATS record1 " +
+                                    "WHERE " +
+                                    "(SELECT COUNT(IDCHAT) FROM USERSCHATS record2 " +
+                                    $"WHERE record1.idchat = record2.idchat AND (IDUSER = {participantsId[0]} OR IDUSER = {participantsId[1]})) = 2 " +
+                                    "AND ROWNUM = 1 AND CHATS.ID = record1.IDCHAT) ";
```
Comment (Hunroll) [inline Cooper.DAO/ChatDAO/ChatDAO.cs]: Is ROWNUM still needed here?

File: Cooper.DAO/MessageDAO/MessageDAO.cs [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+                    messages.Add(message);
+                }
+            }
+
+            return messages;
+        }
+
+        public bool ReadNewMessages(IList<long> messages)
+        {
+            string sqlExpression = String.Format("UPDATE {0} SET ISREAD=\'y\' WHERE ID IN ({1})", table, String.Join(',', messages));
```
Comment (Hunroll) [inline Cooper.DAO/MessageDAO/MessageDAO.cs]: Try to use WhereRequest

**Review by Hunroll — CHANGES_REQUESTED** [review]

File: Cooper.DAO/MessageDAO/MessageDAO.cs [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+
+            return messages;
+        }
+
+        public bool ReadNewMessages(IList<long> messages)
+        {
+            bool areRead = crud.Update(
+                table,
+                new EntityORM() { attributeValue = { { "ISREAD", "\'y\'" } } },
+                new WhereRequest("ID", Operators.In, String.Join(',', messages)));
```
Comment (Hunroll) [inline Cooper.DAO/MessageDAO/MessageDAO.cs]: why String.Join?)

**Review by Hunroll — CHANGES_REQUESTED** [review]

**Review by Hunroll — APPROVED** [review]

**Review by horba — APPROVED** [review]

**Pull request merged**